YouTube Movie Trailer

BrandonCKrueger opened the pull request:
This adds the ability to also include youtube trailer videos with movie selection.  If the current selected movie has an associated youtube trailer it will be displayed on the front page, else it won't be.  Users can also preview their trailer by clicking the 'YouTube' link next to the movie in the list.  Also includes a button to open a new tab to YouTube search with the movie title pre-populated in the search.

<strong>A few screen grabs</strong>:
![My Movie Screen](http://i.imgur.com/WixuFiK.png)

![My Movie Screen Preview](http://i.imgur.com/KNUoGIs.png)

![Home Screen With Movie](http://i.imgur.com/dfngJ4y.jpg)

charwking commented on src/youtube/youtubeValidator.service.js:
```diff
@@ -0,0 +1,35 @@
+                return match[1];
+            }
+            else {
+                return 'dQw4w9WgXcQ';
```
dQw4w9WgXcQ? Sneaky :)

BrandonCKrueger replied:
If people don't submit a video, they get the default ;)

charwking commented:
There's a few _very_ minor things that I'll probably cleanup at some point, but overall this looks pretty good. I don't think the youtube vid on the dashboard will work though. Did you test it? I added you as an admin user so you should be able to Host a Meeting and try it out.

BrandonCKrueger commented on src/dashboard/dashboard.html:
```diff
@@ -2,10 +2,17 @@
 
 <section class="container now-showing-section">
```
To test the dashboard (since I couldn't change movies) I modified this line to be `<section class="container now-showing-section" ng-init="dashboardVm.currentMovie.trailerUrl = ''https://www.youtube.com/watch?v=v2DhIc8s4lY">` which I assumed would be the same as if the video actually had a movie trailer saved.  Now that I'm an admin I'll pull down your changes and re-test it.  If you call out the minor changes you'd like made I'll touch those up too.

ttonyan10 commented:
Yay!!! I'm not in love with the overlay view - is it ok to ditch that and just watch the movie on the page? 

:heart_eyes:

BrandonCKrueger commented:
But that'a the movie theater experience! :smile: 

In all seriousness thought, yeah, I'm fine to change it to whatever.  Maybe display it below the 'Add Movie' button on the left side or something?

charwking commented:
The reason I don't think the dashboard part will work is because I [fail to save the full movie](https://github.com/charwking/movie-club/blob/master/src/hostMeeting/hostMeeting.controller.js#L55) when selecting a new one, so I don't think `dashboardVm.currentMovie.trailerUrl` will ever get a value.

Minor things that you can clean up if you'd like:
* Use the object syntax in the `ng-class` directives. I think this just reads cleaner. Example: `ng-class="{'display-flex': dashboardVm.currentMovie.trailerUrl}"`
* I'm not sure it's necessary to [check that the url is defined and a non-empty string](https://github.com/charwking/movie-club/pull/122/files#diff-6852da8c1de4ce6278a60f60c9cb4dc9R61) prior to calling `validateYoutubeUrl`. Won't the validator handle this?
* There seems to be an extraneous `Movie` attribute in [the movie name input tag](https://github.com/charwking/movie-club/pull/122/files#diff-f2a3841eef608d4ba2c72a83a9bdc6aeR6).
* I think if you avoid [specifying the protocol in the URLs](https://github.com/charwking/movie-club/pull/122/files#diff-f2a3841eef608d4ba2c72a83a9bdc6aeR9) then the browser will reuse whatever protocol is in use by the current page.
* The code in [youtubeValidator.validateYoutubeUrl](https://github.com/charwking/movie-club/pull/122/files#diff-dbd1c144ce3bd9b55bbd5813f66fe8bbR15) and [youtubeValidator.youtubeIdParser](https://github.com/charwking/movie-club/pull/122/files#diff-dbd1c144ce3bd9b55bbd5813f66fe8bbR23) is very, very similar. I'd love to see this DRYed up a little. Also, I might rename `youtubeValidator` since it's doing more than validation (see [youtubeIdParser](https://github.com/charwking/movie-club/pull/122/files#diff-dbd1c144ce3bd9b55bbd5813f66fe8bbR23)). 

Things that I'd eventually change just because I'm OCD (that you can totally ignore):
* Break out the [new config function in app.js](https://github.com/charwking/movie-club/pull/122/files#diff-bd9c9dcd314f2d7df52935b3a6a4d504R15) instead of defining it inline and indent the array values.
* Add a line break before [the watchTrailer function](https://github.com/charwking/movie-club/pull/122/files#diff-6852da8c1de4ce6278a60f60c9cb4dc9R24).
* This is a JavaScript pattern I see frequently for [setting default values on variables](https://github.com/charwking/movie-club/pull/122/files#diff-f589f37aa0fb1cadeab0b8b7adf2ed23R18) that I kind of like: `$scope.width = $scope.width || 560;`
* I'd probably break [this template](https://github.com/charwking/movie-club/pull/122/files#diff-f589f37aa0fb1cadeab0b8b7adf2ed23R29) out into an html file.
* I'm trying to add tests with all the new stuff I add. I've set a poor example with the initial code, so I would accept this PR without the tests, but you're welcome to add some if you want. :)
* Please add your name to the [readme under the contributors section](https://github.com/charwking/movie-club/blob/master/README.md#contributors) too!

BrandonCKrueger commented:
**Major Fix**
You were correct, the trailer was not being saved/retrieved.  Now that I have access as an admin I was able to change it so it saves and gets the trailer URL.

**Minor Fixes**
* Changed `ng-Class`
* Hey man, that `Movie` attribute was there when I got here
* I left it as https:// solely because that's what youtube uses, and since we're opening a blank window I figure why not just use their protocol instead of possibly sending users to http://youtube and having it redirect to https://
* After reading your message and thinking about how it works I rewrote that service so it's only one method since returning the youtube ID is the same as returning true (as far a conditional is concerned)

**OCD Fixes**
* I think I meant to do that as I tested what exactly needed to be white listed on the configuration but kept pushing it off and forgot about it
* Space added before the watchTrailer function
* Removed the ternary style for setting default height and width -- I like the suggested style much more!


**TODO**
* Move the youtube template into it's own HTML file
* Add unit tests

charwking commented:
I'll take a closer look this evening, but after a quick glance I think this is totally mergable. You, or I, or someone else can submit follow on PRs for the remaining TODO items. Excellent work @BrandonCKrueger!

charwking commented:
I'm merging this to master, which will publish to the dev site (prepend 'dev.' to the prod site host). I'll play with it a little there and if all looks good, I'll promote to PROD after a bit.

charwking commented:
Welcome to the ~~cool~~ nerd kids club @BrandonCKrueger :)
